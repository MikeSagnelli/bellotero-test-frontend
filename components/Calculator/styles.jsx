import styled from 'styled-components';
import { textColor, primaryColor, whiteColor } from '../../static/styleConstants';

export const Container = styled.div`
  display: flex;
  justify-content: space-between;
`;

export const InnerContainer = styled.div`
  width: 50%;
  margin-top: -30%;
`;

export const Description = styled.p`
  margin-top: 28px;
  font-family: 'Roboto', sans-serif;
  width: 349px;
  font-size: 16px;
  line-height: 1.5;
  color: ${textColor};
`;

export const TitleContainer = styled.div`
  width: fit-content;
  margin-bottom: 4px;
  display: flex;
  height: 56px;
  background-color: ${primaryColor};
`;

export const TitleText = styled.h1`
  margin: 8px 4px;
  font-family: 'Roboto', sans-serif;
  font-size: 36px;
  font-weight: 900;
  line-height: 1.11;
  letter-spacing: 0.8px;
  color: ${whiteColor};
`;

export const Dollar = styled.h2`
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  margin: 0;
  margin-right: 8px;
  font-family: 'Roboto', sans-serif;
  font-size: 36px;
  font-weight: 500;
  line-height: 1.5;
  letter-spacing: 0.8px;
  color: ${primaryColor};
`;

export const Number = styled.h1`
  max-width: 50%;
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  margin: 0;
  font-family: 'Roboto', sans-serif;
  font-size: 72px;
  font-weight: 500;
  line-height: 1;
  letter-spacing: 1px;
  text-align: right;
  color: ${primaryColor};
`;

export const Label = styled.b`
  display: flex;
  justify-content: flex-end;
  width: 100%;
  font-family: 'Roboto', sans-serif;
  font-size: 14px;
  font-weight: bold;
  color: ${textColor};
`;

export const LabelContainer = styled.div`
  margin-top: 8px;
  display: flex;
  width: 100%;
  justify-content: space-between;
`;

export const ValuesContainer = styled.div`
  display: flex;
  width: 100%;
  justify-content: space-between;
`;

export const ValueContainer = styled.div`
  display: flex;
  justify-content: center;
  width: 50%;
  margin-top: 40px;
`;

export const InputContainer = styled.div`
  display: flex;
  justify-content: space-between;
`;

export const RangeContainer = styled.div`
  width: 100%;
  height: fit-content;
  margin-bottom: 46px;
`;

const NumberInput = styled.input`
  text-align: right;
  border-radius: 4px;
  height: 52px;
  max-width: 50%;
  border: solid 1px #d6dcff;
  background-color: ${whiteColor};
  font-family: 'Roboto', sans-serif;
  font-size: 36px;
  font-weight: 500;
  line-height: 1.11;
  letter-spacing: 0.8px;
  color: ${textColor};
  padding-right: 12px;
`;

export const Money = styled(NumberInput)`
  background: url(/static/images/dollar.svg) no-repeat ${whiteColor} 7px 7px;
  padding-left: 30px;
  width: 190px;
`;

export const Unit = styled(NumberInput)`
  width: 67px;
`;

export const SliderLabel = styled.b`
  width: 157px;
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  font-family: 'Roboto', sans-serif;
  font-size: 14px;
  font-weight: bold;
  color: ${textColor};
`;

export const Range = styled.input`
  &[type='range'] {
    -webkit-appearance: none;
    width: 100%;
    margin-top: 22px;
  }

  &[type='range']:focus {
    outline: none;
  }

  &[type='range']::-webkit-slider-runnable-track {
    width: 100%;
    height: 4px;
    cursor: pointer;
    box-shadow: 0px 0px 1.4px rgba(0, 0, 0, 0), 0px 0px 0px rgba(13, 13, 13, 0);
    background: #f0f2ff;
    border-radius: 25px;
    border: 0px solid rgba(0, 0, 0, 0);
  }

  &[type='range']::-webkit-slider-thumb {
    box-shadow: 2px 2px 8px rgba(7, 30, 179, 0.24), 0px 0px 2px rgba(8, 34, 204, 0.24);
    border: 5px solid #071eb3;
    height: 16px;
    width: 16px;
    border-radius: 50px;
    background: #ffffff;
    cursor: pointer;
    -webkit-appearance: none;
    margin-top: -6px;
  }
`;
