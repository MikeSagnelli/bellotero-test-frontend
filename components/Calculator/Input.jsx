import React from 'react';
import PropTypes from 'prop-types';
import { InputContainer, SliderLabel, Unit, Money, RangeContainer, Range } from './styles';

class Input extends React.PureComponent {
  handleChange(e) {
    const { onChange } = this.props;
    onChange(e.target.name, e.target.value);
  }

  render() {
    const { value, label, type, name, min, max, step, id } = this.props;
    return (
      <React.Fragment>
        <InputContainer>
          <SliderLabel>{label}</SliderLabel>
          {type === 'money' && <Money type="text" value={value} disabled />}
          {type !== 'money' && <Unit type="text" value={value} disabled />}
        </InputContainer>
        <RangeContainer>
          <Range
            type="range"
            id={id}
            min={min}
            max={max}
            step={step}
            value={value}
            onChange={e => this.handleChange(e)}
            name={name}
          />
        </RangeContainer>
      </React.Fragment>
    );
  }
}

Input.displayName = 'Input';

Input.propTypes = {
  value: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,
  min: PropTypes.string.isRequired,
  max: PropTypes.string.isRequired,
  step: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  type: PropTypes.string
};

Input.defaultProps = {
  type: ''
};

export default Input;
