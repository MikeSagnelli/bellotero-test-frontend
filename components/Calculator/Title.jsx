import React from 'react';
import PropTypes from 'prop-types';
import { TitleContainer, TitleText } from './styles';

const Title = ({ title, id }) => (
  <TitleContainer>
    <TitleText id={id}>{title}</TitleText>
  </TitleContainer>
);

export default Title;

Title.displayName = 'Title';

Title.propTypes = {
  title: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired
};
