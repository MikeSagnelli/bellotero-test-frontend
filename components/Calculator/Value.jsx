import React from 'react';
import PropTypes from 'prop-types';
import { Dollar, Number, ValueContainer } from './styles';

const Value = ({ value, id }) => (
  <React.Fragment>
    <ValueContainer>
      <Dollar>$</Dollar>
      <Number id={id}>{value}</Number>
    </ValueContainer>
  </React.Fragment>
);

Value.displayName = 'Value';

Value.propTypes = {
  value: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired
};

export default Value;
