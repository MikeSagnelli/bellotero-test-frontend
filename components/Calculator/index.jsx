import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  Container,
  InnerContainer,
  ValuesContainer,
  LabelContainer,
  Label,
  Description
} from './styles';
import Title from './Title';
import Value from './Value';
import Input from './Input';

class Calculator extends React.PureComponent {
  state = {
    employees: '0',
    costs: '0'
  };

  onChange = (name, value) => {
    this.setState({ [[name]]: value });
  };

  estimateCostsSavings = () => {
    const { costs } = this.state;
    const cost = parseInt(costs, 10);
    const result = cost * 0.3;
    return parseFloat(result, 10).toFixed(2);
  };

  estimateAnnualSavings = () => {
    const { employees } = this.state;
    const nEmployees = parseInt(employees, 10);
    const estimatedCostSavings = this.estimateCostsSavings();
    const result = nEmployees * 1337 + estimatedCostSavings;
    return parseFloat(result, 10).toFixed(2);
  };

  render() {
    const { configuratorData } = this.props;
    const { employees, costs } = this.state;

    const motto = configuratorData.calculator.title.slice(0, 14);
    const companyName = configuratorData.calculator.title.slice(15);
    const estimatedCostSaving = this.estimateCostsSavings();
    const estimatedAnnualSaving = this.estimateAnnualSavings();

    return (
      <Container>
        <InnerContainer>
          <Title id="title-1" title={motto} />
          <Title id="title-2" title={companyName} />
          <Description id="description">{configuratorData.calculator.description}</Description>
        </InnerContainer>
        <InnerContainer>
          <Input
            id="costs"
            name="costs"
            value={costs}
            min="0"
            max="100"
            step="0.1"
            onChange={this.onChange}
            type="money"
            label="Monthly ingredient spending"
          />
          <Input
            id="employees"
            name="employees"
            value={employees}
            min="0"
            max="10"
            step="1"
            onChange={this.onChange}
            label="Full-time employees that process invoices"
          />
          <ValuesContainer>
            <Value id="cost" value={estimatedCostSaving} />
            <Value id="annual" value={estimatedAnnualSaving} />
          </ValuesContainer>
          <LabelContainer>
            <Label>Estimated cost food savings</Label>
            <Label>Your estimated annual savings</Label>
          </LabelContainer>
        </InnerContainer>
      </Container>
    );
  }
}

Calculator.displayName = 'Calculator';

Calculator.propTypes = {
  configuratorData: PropTypes.shape({
    calculator: PropTypes.shape({
      title: PropTypes.string,
      description: PropTypes.string
    })
  })
};

Calculator.defaultProps = {
  configuratorData: null
};

export default connect(state => state)(Calculator);
