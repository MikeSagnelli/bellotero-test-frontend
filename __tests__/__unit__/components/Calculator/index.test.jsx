import React from 'react';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import wait from 'waait';
import Calculator from '../../../../components/Calculator';

const mockStore = configureStore();

const estimateCostsSavings = costs => {
  const cost = parseInt(costs, 10);
  const result = cost * 0.3;
  return parseFloat(result, 10).toFixed(2);
};

const estimateAnnualSavings = (employees, costs) => {
  const nEmployees = parseInt(employees, 10);
  const estimatedCostSavings = estimateCostsSavings(costs);
  const result = nEmployees * 1337 + estimatedCostSavings;
  return parseFloat(result, 10).toFixed(2);
};

const initialState = {
  configuratorData: {
    calculator: {
      title: 'Save more with Bellotero.io',
      description:
        'With Bellotero.io you save time and money make real-time decisions that boost your business and your bottom line. Get less wrongfully blocked payments, save time on bookkeeping and no need to worry about safety.'
    }
  }
};

describe('Calculator', () => {
  it('renders calculator', () => {
    const component = shallow(<Calculator />);
    expect(component).toMatchSnapshot();
  });

  it('renders title and description correctly', async () => {
    const store = mockStore(initialState);

    const component = mount(
      <Provider store={store}>
        <Calculator />
      </Provider>
    );

    await wait(0);
    component.update();

    const title1 = component.find('h1#title-1').text();
    const title2 = component.find('h1#title-2').text();
    const description = component.find('p#description').text();
    expect(title1).toBe(initialState.configuratorData.calculator.title.slice(0, 14));
    expect(title2).toBe(initialState.configuratorData.calculator.title.slice(15));
    expect(description).toBe(initialState.configuratorData.calculator.description);

    component.unmount();
  });

  it('renders correct calculus when changing input range', async () => {
    const store = mockStore(initialState);
    const event1 = { target: { name: 'costs', value: '10' } };
    const event2 = { target: { name: 'employees', value: '2' } };

    const component = mount(
      <Provider store={store}>
        <Calculator />
      </Provider>
    );

    await wait(0);
    component.update();

    const range1 = component.find('input#costs');
    const range2 = component.find('input#employees');

    range1.simulate('change', event1);
    range2.simulate('change', event2);

    await wait(0);
    component.update();

    const cost = component.find('h1#cost').text();
    const annual = component.find('h1#annual').text();

    expect(cost).toBe(estimateCostsSavings('10'));
    expect(annual).toBe(estimateAnnualSavings('2', '10'));

    component.unmount();
  });
});
