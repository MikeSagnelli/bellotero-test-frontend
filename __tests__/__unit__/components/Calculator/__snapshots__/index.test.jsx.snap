// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator renders calculator 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
